Support EnumerableShardingSpec in ShardedTensor.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59776 Add RRef support to ShardedTensor.
* **#59061 Support EnumerableShardingSpec in ShardedTensor.**

Overall Design: https://github.com/pytorch/pytorch/issues/55207

This PR builds upon https://github.com/pytorch/pytorch/pull/58517 and
https://github.com/pytorch/pytorch/pull/57409 to support creating a
ShardedTensor using EnumerableShardingSpec.

Differential Revision: [D28734551](https://our.internmc.facebook.com/intern/diff/D28734551/)

pritamdamania committed:
Support EnumerableShardingSpec in ShardedTensor.

Overall Design: https://github.com/pytorch/pytorch/issues/55207

This PR builds upon https://github.com/pytorch/pytorch/pull/58517 and
https://github.com/pytorch/pytorch/pull/57409 to support creating a
ShardedTensor using EnumerableShardingSpec.

Differential Revision: [D28734551](https://our.internmc.facebook.com/intern/diff/D28734551/)

[ghstack-poisoned]

SciPioneer commented on test/distributed/_sharded_tensor/test_sharded_tensor.py:
```diff
@@ -283,3 +289,269 @@ def test_insufficient_sharding_dims(self):
+
+    @skip_if_lt_x_gpu(4)
+    @requires_nccl()
+    def test_multiple_local_shards(self):
```
Can you add a few more test cases to answer the following questions?
1) Do we allow overlap among different shards?
2) Do we allow the union of all the shards smaller than the input large tensor?

pritamdamania87 replied:
I don't think we support any overlaps amongst any shards at the moment. All the shards should be mutually exclusive.

SciPioneer replied:
Can we add a note at somewhere to declare this or shall we add a check to fail such invalid input?

pritamdamania87 replied:
Makes sense I will update the docs to add this, regarding the check we have the check and a test for this as well: https://github.com/pytorch/pytorch/blob/master/test/distributed/_sharding_spec/test_sharding_spec.py#L158.

pritamdamania committed:
Update on "Support EnumerableShardingSpec in ShardedTensor."


Overall Design: https://github.com/pytorch/pytorch/issues/55207

This PR builds upon https://github.com/pytorch/pytorch/pull/58517 and
https://github.com/pytorch/pytorch/pull/57409 to support creating a
ShardedTensor using EnumerableShardingSpec.

Differential Revision: [D28734551](https://our.internmc.facebook.com/intern/diff/D28734551/)

[ghstack-poisoned]

pritamdamania committed:
Update on "Support EnumerableShardingSpec in ShardedTensor."


Overall Design: https://github.com/pytorch/pytorch/issues/55207

This PR builds upon https://github.com/pytorch/pytorch/pull/58517 and
https://github.com/pytorch/pytorch/pull/57409 to support creating a
ShardedTensor using EnumerableShardingSpec.

Differential Revision: [D28734551](https://our.internmc.facebook.com/intern/diff/D28734551/)

[ghstack-poisoned]

cbalioglu commented on torch/distributed/_sharding_spec/api.py:
```diff
@@ -208,12 +208,12 @@ def _check_shard_pair_overlap(shard1: ShardMetadata, shard2: ShardMetadata):
         return True
 
-    def check_tensor(self, tensor: torch.Tensor) -> None:
+    def check_tensor(self, tensor_dims) -> None:
```
Nit: `tensor_dims` can be type annotated as `tensor_dims: Sequence[int]`.
